File: nine-language-version/lessons/06-recursion/recursion.pl
% Lesson 6: Recursion in Prolog
%
% Prolog is a logic programming language where recursion is FUNDAMENTAL.
% All iteration is done through recursion. Prolog has tail call optimization.
% Rules are defined recursively with base cases and recursive cases.

% ====================
% 1. Simple Recursion
% ====================

% Factorial
factorial(0, 1).  % Base case
factorial(N, Result) :-
    N > 0,
    N1 is N - 1,
    factorial(N1, SubResult),
    Result is N * SubResult.

% Tail-recursive factorial (more efficient)
factorial_tail(N, Result) :-
    factorial_tail(N, 1, Result).

factorial_tail(0, Acc, Acc).  % Base case
factorial_tail(N, Acc, Result) :-
    N > 0,
    N1 is N - 1,
    Acc1 is N * Acc,
    factorial_tail(N1, Acc1, Result).

% ====================
% 2. Fibonacci
% ====================

% Fibonacci (inefficient)
fibonacci(0, 0).
fibonacci(1, 1).
fibonacci(N, Result) :-
    N > 1,
    N1 is N - 1,
    N2 is N - 2,
    fibonacci(N1, F1),
    fibonacci(N2, F2),
    Result is F1 + F2.

% ====================
% 3. List Recursion
% ====================

% Sum of list
sum_list([], 0).  % Base case: empty list sums to 0
sum_list([Head|Tail], Sum) :-
    sum_list(Tail, SubSum),
    Sum is Head + SubSum.

% Length of list
list_length([], 0).
list_length([_|Tail], Length) :-
    list_length(Tail, SubLength),
    Length is SubLength + 1.

% Reverse list
reverse_list([], []).
reverse_list([Head|Tail], Reversed) :-
    reverse_list(Tail, ReversedTail),
    append(ReversedTail, [Head], Reversed).

% Tail-recursive reverse (more efficient)
reverse_tail(List, Reversed) :-
    reverse_tail(List, [], Reversed).

reverse_tail([], Acc, Acc).
reverse_tail([Head|Tail], Acc, Reversed) :-
    reverse_tail(Tail, [Head|Acc], Reversed).

% Maximum element
max_element([X], X).  % Base case: single element
max_element([Head|Tail], Max) :-
    max_element(Tail, TailMax),
    (Head > TailMax -> Max = Head ; Max = TailMax).

% ====================
% 4. List Membership
% ====================

% Check if element is in list
member_recursive(X, [X|_]).  % Base case: X is head
member_recursive(X, [_|Tail]) :-
    member_recursive(X, Tail).  % Recursive: check tail

% ====================
% 5. List Append
% ====================

% Append two lists
append_recursive([], List, List).  % Base case
append_recursive([Head|Tail], List2, [Head|Result]) :-
    append_recursive(Tail, List2, Result).

% ====================
% 6. Quicksort
% ====================

% Quicksort
quicksort([], []).  % Base case: empty list
quicksort([Pivot|Rest], Sorted) :-
    partition(Pivot, Rest, Smaller, Larger),
    quicksort(Smaller, SortedSmaller),
    quicksort(Larger, SortedLarger),
    append(SortedSmaller, [Pivot|SortedLarger], Sorted).

% Partition list around pivot
partition(_, [], [], []).
partition(Pivot, [X|Rest], [X|Smaller], Larger) :-
    X < Pivot,
    partition(Pivot, Rest, Smaller, Larger).
partition(Pivot, [X|Rest], Smaller, [X|Larger]) :-
    X >= Pivot,
    partition(Pivot, Rest, Smaller, Larger).

% ====================
% 7. Tower of Hanoi
% ====================

% Hanoi solver - prints moves
hanoi(1, Source, Target, _) :-
    format('   Move disk 1 from ~w to ~w~n', [Source, Target]).

hanoi(N, Source, Target, Auxiliary) :-
    N > 1,
    N1 is N - 1,
    hanoi(N1, Source, Auxiliary, Target),
    format('   Move disk ~w from ~w to ~w~n', [N, Source, Target]),
    hanoi(N1, Auxiliary, Target, Source).

% ====================
% 8. Tree Operations
% ====================

% Tree represented as: tree(Value, Left, Right) or nil
% Example: tree(5, tree(3, tree(1, nil, nil), tree(4, nil, nil)), tree(8, nil, tree(9, nil, nil)))

% Tree height
tree_height(nil, 0).
tree_height(tree(_, Left, Right), Height) :-
    tree_height(Left, LeftHeight),
    tree_height(Right, RightHeight),
    MaxHeight is max(LeftHeight, RightHeight),
    Height is MaxHeight + 1.

% Tree sum
tree_sum(nil, 0).
tree_sum(tree(Value, Left, Right), Sum) :-
    tree_sum(Left, LeftSum),
    tree_sum(Right, RightSum),
    Sum is Value + LeftSum + RightSum.

% Inorder traversal
inorder(nil, []).
inorder(tree(Value, Left, Right), Result) :-
    inorder(Left, LeftList),
    inorder(Right, RightList),
    append(LeftList, [Value|RightList], Result).

% ====================
% 9. Mutual Recursion
% ====================

% Check if even/odd using mutual recursion
is_even(0).
is_even(N) :-
    N > 0,
    N1 is N - 1,
    is_odd(N1).

is_odd(N) :-
    N > 0,
    not(is_even(N)).

% ====================
% 10. GCD
% ====================

% Greatest Common Divisor (Euclid's algorithm)
gcd_recursive(A, 0, A).
gcd_recursive(A, B, GCD) :-
    B > 0,
    R is A mod B,
    gcd_recursive(B, R, GCD).

% ====================
% 11. Range Generation
% ====================

% Generate list from Low to High
range(High, High, [High]).
range(Low, High, [Low|Rest]) :-
    Low < High,
    Next is Low + 1,
    range(Next, High, Rest).

% ====================
% 12. Map/Filter
% ====================

% Map: apply predicate to each element
map_list(_, [], []).
map_list(Pred, [Head|Tail], [NewHead|NewTail]) :-
    call(Pred, Head, NewHead),
    map_list(Pred, Tail, NewTail).

% Filter: keep elements that satisfy predicate
filter_list(_, [], []).
filter_list(Pred, [Head|Tail], [Head|FilteredTail]) :-
    call(Pred, Head),
    filter_list(Pred, Tail, FilteredTail).
filter_list(Pred, [_|Tail], FilteredTail) :-
    filter_list(Pred, Tail, FilteredTail).

% ====================
% Test Predicates
% ====================

% Run all tests
run_tests :-
    writeln('=== Recursion Examples in Prolog ==='),
    nl,

    % Factorial
    writeln('1. Factorial:'),
    factorial(5, F1),
    format('   factorial(5) = ~w~n', [F1]),
    factorial_tail(5, F2),
    format('   factorial_tail(5) = ~w~n', [F2]),

    % Fibonacci
    nl,
    writeln('2. Fibonacci:'),
    fibonacci(10, Fib),
    format('   fibonacci(10) = ~w~n', [Fib]),

    % List operations
    nl,
    writeln('3. List Operations:'),
    Numbers = [1, 2, 3, 4, 5],
    sum_list(Numbers, Sum),
    format('   sum_list(~w) = ~w~n', [Numbers, Sum]),
    list_length(Numbers, Len),
    format('   list_length(~w) = ~w~n', [Numbers, Len]),
    reverse_tail(Numbers, Rev),
    format('   reverse_list(~w) = ~w~n', [Numbers, Rev]),
    max_element(Numbers, Max),
    format('   max_element(~w) = ~w~n', [Numbers, Max]),

    % Quicksort
    nl,
    writeln('4. Quicksort:'),
    Unsorted = [3, 6, 8, 10, 1, 2, 1],
    quicksort(Unsorted, Sorted),
    format('   quicksort(~w) = ~w~n', [Unsorted, Sorted]),

    % Tower of Hanoi
    nl,
    writeln('5. Tower of Hanoi (3 disks):'),
    hanoi(3, a, c, b),

    % Tree
    nl,
    writeln('6. Binary Tree:'),
    Tree = tree(5,
                tree(3,
                     tree(1, nil, nil),
                     tree(4, nil, nil)),
                tree(8,
                     nil,
                     tree(9, nil, nil))),
    tree_height(Tree, Height),
    format('   tree_height() = ~w~n', [Height]),
    tree_sum(Tree, TreeSum),
    format('   tree_sum() = ~w~n', [TreeSum]),
    inorder(Tree, Inorder),
    format('   inorder_traversal() = ~w~n', [Inorder]),

    % Mutual recursion
    nl,
    writeln('7. Mutual Recursion:'),
    (is_even(10) -> Even10 = true ; Even10 = false),
    format('   is_even(10) = ~w~n', [Even10]),
    (is_even(7) -> Even7 = true ; Even7 = false),
    format('   is_even(7) = ~w~n', [Even7]),

    % GCD
    nl,
    writeln('8. Greatest Common Divisor:'),
    gcd_recursive(48, 18, GCD1),
    format('   gcd(48, 18) = ~w~n', [GCD1]),
    gcd_recursive(100, 35, GCD2),
    format('   gcd(100, 35) = ~w~n', [GCD2]),

    % Range
    nl,
    writeln('9. Range Generation:'),
    range(1, 10, R),
    format('   range(1, 10) = ~w~n', [R]).

% To run: swipl -s recursion.pl -g run_tests -t halt

File: three-language-version/lessons/06-recursion/recursion.hs
-- Lesson 6: Recursion - Haskell Examples
-- Recursion is the natural way to program in Haskell!

module Main where

import Data.List (intercalate)

-- =============================================================================
-- BASIC RECURSION: Factorial
-- =============================================================================

-- Pattern matching makes the base case explicit
factorial :: Integer -> Integer
factorial 0 = 1                    -- Base case
factorial n = n * factorial (n - 1)  -- Recursive case

-- =============================================================================
-- FIBONACCI SEQUENCE
-- =============================================================================

-- Naive recursive (elegant but slow)
fibonacci :: Int -> Integer
fibonacci 0 = 0
fibonacci 1 = 1
fibonacci n = fibonacci (n - 1) + fibonacci (n - 2)

-- Efficient version using infinite list
fibList :: [Integer]
fibList = 0 : 1 : zipWith (+) fibList (tail fibList)

fibFast :: Int -> Integer
fibFast n = fibList !! n

-- =============================================================================
-- LIST PROCESSING (Haskell's bread and butter!)
-- =============================================================================

-- Sum a list
sumList :: [Int] -> Int
sumList []     = 0              -- Base case: empty list
sumList (x:xs) = x + sumList xs  -- Recursive case: head + sum of tail

-- Length of a list
len :: [a] -> Int
len []     = 0
len (_:xs) = 1 + len xs

-- Check if element is in list
contains :: Eq a => a -> [a] -> Bool
contains _ []     = False
contains target (x:xs)
    | target == x = True
    | otherwise   = contains target xs

-- Reverse a list
reverseList :: [a] -> [a]
reverseList []     = []
reverseList (x:xs) = reverseList xs ++ [x]

-- More efficient tail-recursive reverse
reverseList' :: [a] -> [a]
reverseList' = go []
  where
    go acc []     = acc
    go acc (x:xs) = go (x:acc) xs

-- =============================================================================
-- PATTERN MATCHING WITH RECURSION
-- =============================================================================

-- Take first n elements
takeN :: Int -> [a] -> [a]
takeN 0 _      = []
takeN _ []     = []
takeN n (x:xs) = x : takeN (n-1) xs

-- Drop first n elements
dropN :: Int -> [a] -> [a]
dropN 0 xs     = xs
dropN _ []     = []
dropN n (_:xs) = dropN (n-1) xs

-- Map a function over a list
mapList :: (a -> b) -> [a] -> [b]
mapList _ []     = []
mapList f (x:xs) = f x : mapList f xs

-- Filter elements
filterList :: (a -> Bool) -> [a] -> [a]
filterList _ []     = []
filterList p (x:xs)
    | p x       = x : filterList p xs
    | otherwise = filterList p xs

-- =============================================================================
-- BINARY SEARCH
-- =============================================================================

binarySearch :: Ord a => a -> [a] -> Maybe Int
binarySearch target arr = go 0 (length arr - 1)
  where
    go low high
        | low > high       = Nothing
        | arr !! mid == target = Just mid
        | arr !! mid < target  = go (mid + 1) high
        | otherwise        = go low (mid - 1)
      where
        mid = (low + high) `div` 2

-- =============================================================================
-- TREE STRUCTURES (Natural fit for recursion)
-- =============================================================================

data Tree a = Empty | Node a (Tree a) (Tree a)
    deriving (Show, Eq)

-- Insert into BST
insert :: Ord a => a -> Tree a -> Tree a
insert x Empty = Node x Empty Empty
insert x (Node y left right)
    | x < y     = Node y (insert x left) right
    | otherwise = Node y left (insert x right)

-- In-order traversal
inOrder :: Tree a -> [a]
inOrder Empty = []
inOrder (Node x left right) = inOrder left ++ [x] ++ inOrder right

-- Tree height
treeHeight :: Tree a -> Int
treeHeight Empty = 0
treeHeight (Node _ left right) = 1 + max (treeHeight left) (treeHeight right)

-- =============================================================================
-- ACCUMULATORS AND TAIL RECURSION
-- =============================================================================

-- Tail recursive factorial
factorialTail :: Integer -> Integer
factorialTail n = go n 1
  where
    go 0 acc = acc
    go n acc = go (n - 1) (n * acc)

-- Tail recursive sum
sumTail :: [Int] -> Int
sumTail = go 0
  where
    go acc []     = acc
    go acc (x:xs) = go (acc + x) xs

-- =============================================================================
-- MUTUAL RECURSION
-- =============================================================================

isEven :: Int -> Bool
isEven 0 = True
isEven n = isOdd (n - 1)

isOdd :: Int -> Bool
isOdd 0 = False
isOdd n = isEven (n - 1)

-- =============================================================================
-- FLATTEN NESTED STRUCTURES
-- =============================================================================

data Nested a = Elem a | List [Nested a]
    deriving Show

flatten :: Nested a -> [a]
flatten (Elem x)  = [x]
flatten (List xs) = concatMap flatten xs

-- =============================================================================
-- MAIN
-- =============================================================================

main :: IO ()
main = do
    putStrLn "=== Factorial ==="
    mapM_ (\n -> putStrLn $ show n ++ "! = " ++ show (factorial n)) [0..5]

    putStrLn "\n=== Fibonacci ==="
    putStrLn $ "First 10 Fibonacci: " ++ show (take 10 fibList)

    putStrLn "\n=== List Processing ==="
    let numbers = [1, 2, 3, 4, 5]
    putStrLn $ "sumList " ++ show numbers ++ " = " ++ show (sumList numbers)
    putStrLn $ "len " ++ show numbers ++ " = " ++ show (len numbers)
    putStrLn $ "contains 3 " ++ show numbers ++ " = " ++ show (contains 3 numbers)
    putStrLn $ "contains 9 " ++ show numbers ++ " = " ++ show (contains 9 numbers)
    putStrLn $ "reverseList " ++ show numbers ++ " = " ++ show (reverseList numbers)

    putStrLn "\n=== Binary Search ==="
    let sorted = [1, 3, 5, 7, 9, 11, 13, 15]
    putStrLn $ "binarySearch 7 " ++ show sorted ++ " = " ++ show (binarySearch 7 sorted)
    putStrLn $ "binarySearch 10 " ++ show sorted ++ " = " ++ show (binarySearch 10 sorted)

    putStrLn "\n=== Tree Operations ==="
    let tree = foldr insert Empty [5, 3, 7, 1, 9, 4, 6]
    putStrLn $ "In-order traversal: " ++ show (inOrder tree)
    putStrLn $ "Tree height: " ++ show (treeHeight tree)

    putStrLn "\n=== Tail Recursion ==="
    putStrLn $ "factorialTail 5 = " ++ show (factorialTail 5)
    putStrLn $ "sumTail [1..100] = " ++ show (sumTail [1..100])

    putStrLn "\n=== Mutual Recursion ==="
    mapM_ (\n -> putStrLn $ "isEven " ++ show n ++ " = " ++ show (isEven n) ++
                           ", isOdd " ++ show n ++ " = " ++ show (isOdd n)) [0..4]
